pkg/policy/rule.go

```diff
@@ -41,14 +41,16 @@ func (r *rule) validate() error {
 	return nil
 }
 
-func mergeL4Port(ctx *SearchContext, r api.PortRule, p api.PortProtocol, dir string, proto string, resMap L4PolicyMap) int {
+func mergeL4Port(fromEndpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
+	dir string, proto string, resMap L4PolicyMap) int {
+
 	fmt := p.Port + "/" + proto
 	v, ok := resMap[fmt]
 	if !ok {
-		resMap[fmt] = CreateL4Filter(r, p, dir, proto)
+		resMap[fmt] = CreateL4Filter(fromEndpoints, r, p, dir, proto)
 		return 1
 	}
-	l4Filter := CreateL4Filter(r, p, dir, proto)
+	l4Filter := CreateL4Filter(fromEndpoints, r, p, dir, proto)
```

Comment:
I think the logic should be:
1. Any L7 rule with 1..n combined FromEndpoints must be appended to a list which is maintained for *each* endpoint in the FromEndpoints, e.g. if we have the independent rules: r1 from {a,b} and r2 from {b,c}, then b must be able to do {r1, r2}.
2. Any L7 rule without an explicit combined FromEndpoints must be appended to all lists maintained and also to a wildcard list. We may also maintain the wildcard list only and then append the wildcard list to all maintained lists at the end. E.g. if we have an additional, unconstrained r3 then {a, b, c and d} must be able to do r3.